### PR TITLE
[MAT-1948] SampledData Widget

### PR DESCRIPTION
**Pull request by jkotanchik-SB**
### SampledData Widget

**Composite of quantity, decimal, positive int, and string widgets.** 
- By default, displays required fields and hides optional fields behind a collapse toggle. 
- Added superscript `*` to identify required fields. 

Screen Shot 2020-11-03 at 2 09 00 PM

**Fully Expanded.**

Screen Shot 2020-11-04 at 11 40 17 AM

**To reduce chip clutter, the various sub-values have been separated by pipes.**

Screen Shot 2020-11-03 at 2 09 58 PM

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: [MAT-1948](https://jira.cms.gov/browse/MAT-1948)
- [ ] JIRA ticket links to this PR
- [ ] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] If UI changes have been made, google WAVE plug-in has been executed to ensure no 508 issues were introduced.
- [ ] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [ ] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [ ] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: 
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: 
- [ ] Automated regression test(s) pass

If JIRA tests were used to supplement or replace automated tests:
- [ ] JIRA test links:
- [ ] Justification for using JIRA tests:
- [ ] JIRA tests have been added to sprint

**Reviewer 1:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [ ] JIRA tests have been run and pass
- [ ] You agree with the justification for use of JIRA tests or have provided input on why you disagree

**Reviewer 2:**

Name:
- [ ] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [ ] The tests appropriately test the new code, including edge cases
- [ ] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [ ] JIRA tests have been run and pass
- [ ] You agree with the justification for use of JIRA tests or have provided input on why you disagree

### CODE REVIEW

File: app/assets/javascripts/templates/patient_builder/inputs/sampled_data.hbs [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    <div class="form-group col-md-12 data-view">
+      <div class="control-label">Data</div>
+      <span class="sr-only">data string</span>
+      {{view dataView}}
+    </div>
+  </div>
+  <div class="form-group col-md-12">
+    {{#button trigger="optionalToggled" class="btn btn-primary btn-sm" id="toggle"
+              type="button" data-toggle="collapse" data-target="#optionalElements" data-cy="optional_toogle"
+              aria-expanded="false" aria-controls="optionalElements"}}Show Optional Elements{{/button}}
```
Comment (adongare) [inline app/assets/javascripts/templates/patient_builder/inputs/sampled_data.hbs]: Okay, this is interesting. This will not work when you have more than one resources added that have SampledData type attributes. you have to make ids and targets unique by appending `_@cid`

File: app/assets/javascripts/views/patient_builder/inputs/sampled_data.js.coffee [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  updateValueFromSubviews: ->
+    @value = new cqm.models.SampledData() unless @value?
+    @update view for view in @subviews when view.hasValidValue()
+    @trigger 'valueChanged', @
+
+  updateButton: ->
+    if $('#toggle').text() == 'Show Optional Elements'
+      $('#toggle').text('Hide Optional Elements')
+    else
+      $('#toggle').text('Show Optional Elements')
```
Comment (adongare) [inline app/assets/javascripts/views/patient_builder/inputs/sampled_data.js.coffee]: same here. you probably need to get the target from event instead of hardcoding it.

**Review by ghost — APPROVED** [review]

File: app/assets/javascripts/templates/patient_builder/inputs/sampled_data.hbs (hunk 1 above)
Comment (jkotanchik-SB) [reply app/assets/javascripts/templates/patient_builder/inputs/sampled_data.hbs]: Indeed. I didn't think to test multiple resources in the same patient, and without the cid, the buttons would toggle every open instance of the widget. 

Updated with cid's appended.

File: app/assets/javascripts/views/patient_builder/inputs/sampled_data.js.coffee (hunk 2 above)
Comment (jkotanchik-SB) [reply app/assets/javascripts/views/patient_builder/inputs/sampled_data.js.coffee]: Updated as has prescribed.

**Review by adongare — APPROVED** [review]
